### PR TITLE
Final Exponentiation gadgets

### PR DESCRIPTION
**Pull request by dtebbs**
- Further field operations
- Gadgets for first and last part of final exponentiation

### CODE REVIEW

File: libzecale/circuits/pairing/bls12_377_pairing.tcc [hunk 1]
```diff
@@ -732,6 +732,386 @@ void bls12_377_ate_miller_loop_gadget<ppT>::generate_r1cs_witness()
...
+    } else {
+        initialize_z_pos(pb, annotation_prefix);
+    }
+}
+
+template<typename ppT>
+void bls12_377_exp_by_z_gadget<ppT>::initialize_z_neg(
+    libsnark::protoboard<FieldT> &pb, const std::string &annotation_prefix)
+{
+    const Fp12_2over3over2_variable<FqkT> *res = &_in;
```
Comment (dtebbs) [inline libzecale/circuits/pairing/bls12_377_pairing.tcc:818]: use variable names and descriptions in annotations in this and the other initialization function.

File: libzecale/circuits/fields/fp12_2over3over2_gadgets.tcc [hunk 2]
```diff
@@ -361,6 +412,294 @@ void Fp12_2over3over2_mul_by_024_gadget<Fp12T>::generate_r1cs_witness()
...
+    //   <=> a1*b1 = (result0 - a0*b0) * non_residue.inverse
+    , _a1_times_b1(
+          pb,
+          A._c1,
+          B._c1,
+          fp6_mul_by_non_residue_inverse<Fp12T>(
+              pb,
+              _result._c0 - _a0_times_b0._result,
+              FMT(annotation_prefix, " a1*b1")),
+          FMT(annotation_prefix, " _a1_times_b1"))
```
Comment (AntoineRondelet) [inline libzecale/circuits/fields/fp12_2over3over2_gadgets.tcc:444]: Since you're using Karatsuba here, let's switch the annotations to `v0` (resp. `v1`) for `a0*b0` (resp. `a1*b1`). See: https://github.com/clearmatics/zecale/pull/17#discussion_r450927132

File: libzecale/circuits/fields/fp6_3over2_gadgets.tcc [hunk 3]
```diff
@@ -83,6 +112,37 @@ Fp6_3over2_variable<Fp6T> Fp6_3over2_variable<Fp6T>::operator+(
...
+template<typename Fp6T>
+Fp6_3over2_variable<Fp6T> Fp6_3over2_variable<Fp6T>::operator-(
+    const Fp6_3over2_variable<Fp6T> &other) const
+{
+    return Fp6_3over2_variable<Fp6T>(
+        this->pb,
+        _c0 - other._c0,
+        _c1 - other._c1,
+        _c2 - other._c2,
+        FMT(this->annotation_prefix, " - a"));
```
Comment (AntoineRondelet) [inline libzecale/circuits/fields/fp6_3over2_gadgets.tcc:124]: What is `a` in the annotation? Shouldn't it be `-other`?
Moreover, why did you add a space between `-` and `a`? the space is the annotation separator, right?

File: libzecale/circuits/pairing/bls12_377_pairing.hpp [hunk 4]
```diff
@@ -325,6 +325,119 @@ class bls12_377_ate_miller_loop_gadget : public libsnark::gadget<libff::Fr<ppT>>
...
+class bls12_377_final_exp_first_part_gadget
+    : public libsnark::gadget<libff::Fr<ppT>>
+{
+public:
+    using FieldT = libff::Fr<ppT>;
+    using FqkT = libff::Fqk<libsnark::other_curve<ppT>>;
+
+    // Follows the implementation used in
+    // libff::bls12_377_final_exponentiation_first_chunk() (see
+    // libff/algebra/curves/bls12_377/bls12_377_pairing.cpp), which in turn
```
Comment (AntoineRondelet) [inline libzecale/circuits/pairing/bls12_377_pairing.hpp:338]: `libff/algebra/curves/bls12_377/bls12_377_pairing.cpp` -> `clearmatics/libff/algebra/curves/bls12_377/bls12_377_pairing.cpp` (the scipr-lab `libff` library does not yet contain this code)

File: libzecale/circuits/pairing/bls12_377_pairing.hpp [hunk 5]
```diff
@@ -325,6 +325,119 @@ class bls12_377_ate_miller_loop_gadget : public libsnark::gadget<libff::Fr<ppT>>
...
+        const Fp12_2over3over2_variable<FqkT> &result,
+        const std::string &annotation_prefix);
+
+    const Fp12_2over3over2_variable<FqkT> &result() const;
+    void generate_r1cs_constraints();
+    void generate_r1cs_witness();
+
+private:
+    void initialize_z_neg(libsnark::protoboard<FieldT> &, const std::string &);
+    void initialize_z_pos(libsnark::protoboard<FieldT> &, const std::string &);
```
Comment (AntoineRondelet) [inline libzecale/circuits/pairing/bls12_377_pairing.hpp:393]: What about adding arg names here?, i.e. `pb` and `str`

File: libzecale/tests/circuits/bls12_377_pairing_test.cpp [hunk 6]
```diff
@@ -368,6 +368,150 @@ TEST(BLS12_377_PairingTest, MillerLoopGadgetTest)
...
+    // Circuit with final exponentiation first part gadget
+    libsnark::protoboard<FieldT> pb;
+    libzecale::Fp12_2over3over2_variable<FqkT> a_var(pb, "a");
+    libzecale::Fp12_2over3over2_variable<FqkT> final_exp_first_part_var(
+        pb, "final_exp_first_part");
+    const size_t num_primary_inputs = pb.num_inputs();
+    pb.set_input_sizes(num_primary_inputs);
+    libzecale::bls12_377_final_exp_first_part_gadget<wpp>
+        final_exp_first_part_gadget(
+            pb, a_var, final_exp_first_part_var, "final exp first part");
```
Comment (AntoineRondelet) [inline libzecale/tests/circuits/bls12_377_pairing_test.cpp:400]: space is the annotation separator. `final exp first part` -> `compute_final_exp_first_part` (the variable `final_exp_first_part_var` already as the annotation `"final_exp_first_part"`)

File: libzecale/tests/circuits/bls12_377_pairing_test.cpp [hunk 7]
```diff
@@ -368,6 +368,150 @@ TEST(BLS12_377_PairingTest, MillerLoopGadgetTest)
...
+    // Circuit with final exponentiation last part gadget
+    libsnark::protoboard<FieldT> pb;
+    libzecale::Fp12_2over3over2_variable<FqkT> a_var(pb, "a");
+    libzecale::Fp12_2over3over2_variable<FqkT> final_exp_last_part_var(
+        pb, "final_exp_last_part");
+    const size_t num_primary_inputs = pb.num_inputs();
+    pb.set_input_sizes(num_primary_inputs);
+    libzecale::bls12_377_final_exp_last_part_gadget<wpp>
+        final_exp_last_part_gadget(
+            pb, a_var, final_exp_last_part_var, "final exp last part");
```
Comment (AntoineRondelet) [inline libzecale/tests/circuits/bls12_377_pairing_test.cpp:497]: Same as above `final exp last part` -> `compute_final_exp_last_part`

File: libzecale/tests/circuits/fp12_2over3over2_gadgets_test.cpp [hunk 8]
```diff
@@ -169,6 +234,225 @@ TEST(Fp12_2over3over2_Test, MulBy024GadgetTest)
...
+            Fp2T(FieldT("21"), FieldT("22")),
+            Fp2T(FieldT("23"), FieldT("24")),
+            Fp2T(FieldT("25"), FieldT("26"))));
+    const Fp12T u = libff::bls12_377_final_exponentiation_first_chunk(a);
+    const Fp12T u_squared = u.cyclotomic_squared();
+
+    // Inversion in a circuit
+    libsnark::protoboard<FieldT> pb;
+    libzecale::Fp12_2over3over2_variable<Fp12T> u_var(pb, "u");
+    libzecale::Fp12_2over3over2_variable<Fp12T> u_squared_var(pb, "u squared");
```
Comment (AntoineRondelet) [inline libzecale/tests/circuits/fp12_2over3over2_gadgets_test.cpp:391]: `u squared` -> `u_squared`

File: libzecale/circuits/pairing/bls12_377_pairing.hpp (hunk 4 above)
Comment (AntoineRondelet) [reply libzecale/circuits/pairing/bls12_377_pairing.hpp:338]: Note: This is applicable to other places in this PR

File: libzecale/circuits/pairing/bls12_377_pairing.tcc (hunk 1 above)
Comment (dtebbs) [reply libzecale/circuits/pairing/bls12_377_pairing.tcc:818]: addressed

File: libzecale/circuits/fields/fp12_2over3over2_gadgets.hpp [hunk 9]
```diff
@@ -32,17 +33,45 @@ class Fp12_2over3over2_variable : public libsnark::gadget<typename Fp12T::my_Fp>
...
+    Fp12_2over3over2_variable<Fp12T> operator*(const Fp2T &fp2_const) const;
+    Fp12_2over3over2_variable<Fp12T> frobenius_map(size_t power) const;
+    Fp12_2over3over2_variable<Fp12T> unitary_inverse() const;
+    void evaluate() const;
     void generate_r1cs_witness(const Fp12T &v);
+    Fp12T get_element() const;
 };
 
+/// Multiply an element in Fq6 by Fq12::non_residue. Let c = c0 + c1 * v +
+/// c2.v^2 be an element of Fq6T, where v is a root of:
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply libzecale/circuits/fields/fp12_2over3over2_gadgets.hpp]: changed

File: libzecale/circuits/fields/fp12_2over3over2_gadgets.hpp [hunk 10]
```diff
@@ -32,17 +33,45 @@ class Fp12_2over3over2_variable : public libsnark::gadget<typename Fp12T::my_Fp>
...
+/// circuit since Fp6_3over2_variable::operator*(const Fp6T &)
+/// (multiplication by a constant) can be implemented as linear
+/// combinations.
+template<typename Fp12T>
+Fp6_3over2_variable<typename Fp12T::my_Fp6> fp6_mul_by_non_residue(
+    libsnark::protoboard<typename Fp12T::my_Fp> &pb,
+    const Fp6_3over2_variable<typename Fp12T::my_Fp6> &c,
+    const std::string &annotation_prefix);
+
+/// Let c = c0 + c1.v + c2.v^2 be an element of Fq6T, where v is a root of:
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply libzecale/circuits/fields/fp12_2over3over2_gadgets.hpp]: changed

File: libzecale/circuits/fields/fp12_2over3over2_gadgets.hpp [hunk 11]
```diff
@@ -32,17 +33,45 @@ class Fp12_2over3over2_variable : public libsnark::gadget<typename Fp12T::my_Fp>
...
+///   v^3 - Fq6::non_residue
+/// and v is used as Fp12::non_residue.
+/// Return c * v^{-1} ( = c * Fp12::non_residue^{-1})
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply libzecale/circuits/fields/fp12_2over3over2_gadgets.hpp]: removed space

File: libzecale/circuits/fields/fp12_2over3over2_gadgets.tcc (hunk 2 above)
Comment (dtebbs) [reply libzecale/circuits/fields/fp12_2over3over2_gadgets.tcc]: switched to v0, v1

File: libzecale/circuits/fields/fp6_3over2_gadgets.tcc (hunk 3 above)
Comment (dtebbs) [reply libzecale/circuits/fields/fp6_3over2_gadgets.tcc]: changed

File: libzecale/circuits/pairing/bls12_377_pairing.hpp (hunk 4 above)
Comment (dtebbs) [reply libzecale/circuits/pairing/bls12_377_pairing.hpp]: changed references to clearmatics/libff

File: libzecale/circuits/pairing/bls12_377_pairing.hpp (hunk 5 above)
Comment (dtebbs) [reply libzecale/circuits/pairing/bls12_377_pairing.hpp]: added args

File: libzecale/tests/circuits/bls12_377_pairing_test.cpp (hunk 6 above)
Comment (dtebbs) [reply libzecale/tests/circuits/bls12_377_pairing_test.cpp]: chaged

File: libzecale/tests/circuits/bls12_377_pairing_test.cpp (hunk 7 above)
Comment (dtebbs) [reply libzecale/tests/circuits/bls12_377_pairing_test.cpp]: changed

File: libzecale/tests/circuits/fp12_2over3over2_gadgets_test.cpp (hunk 8 above)
Comment (dtebbs) [reply libzecale/tests/circuits/fp12_2over3over2_gadgets_test.cpp]: changed

File: libzecale/circuits/fields/fp12_2over3over2_gadgets.tcc (hunk 2 above)
Comment (AntoineRondelet) [reply libzecale/circuits/fields/fp12_2over3over2_gadgets.tcc]: So you renamed the gadget and kept the wire name, right? I'd lean toward naming the output wire of the circuit `v1`.
We have already talked about the need to iterate/polish the annotations as part of a PR once all these a merged. Raising this here not to forget, but feel free to defer this to a later PR

Comment (AntoineRondelet) [reply libzecale/circuits/fields/fp12_2over3over2_gadgets.tcc]: Opened a ticket for that here: https://github.com/clearmatics/zecale/issues/20

File: libzecale/circuits/pairing/bls12_377_pairing.hpp (hunk 4 above)
Comment (AntoineRondelet) [reply libzecale/circuits/pairing/bls12_377_pairing.hpp]: Just as a quick note, remember that `libff` contains a `libff` repo that contains the source of the library. Strictly speaking, we need `clearmatics/libff/libff`, but that's a detail

File: libzecale/tests/circuits/bls12_377_pairing_test.cpp (hunk 6 above)
Comment (AntoineRondelet) [reply libzecale/tests/circuits/bls12_377_pairing_test.cpp]: This may be another one to rename, i.e. we may want to rename `final_exp_first_part_gadget` as `compute_final_exp_first_part` - which would be in line with the `compute_v1/0` names for the gadgets used in Karatsuba and would also be in line with the annotation used here.
I'll link this comment to the associated ticket on annotations.

File: libzecale/circuits/fields/fp12_2over3over2_gadgets.tcc (hunk 2 above)
Comment (dtebbs) [reply libzecale/circuits/fields/fp12_2over3over2_gadgets.tcc]: Yes, that makes sense for the future.  My thinking is that the gadgets will eventually be renamed to `compute_<name>` in a later pass, at which point it will make sense to change the annotations to match, and then use names like `v0` etc for the transient variables that hold those values.  For now I thought it best to consistently make the annotations match the member names.

File: libzecale/circuits/pairing/bls12_377_pairing.hpp (hunk 4 above)
Comment (dtebbs) [reply libzecale/circuits/pairing/bls12_377_pairing.hpp]: Good point.  I'll fix this as part of #19
